### DISCUSSION TITLE
Is there a way to assign different UID:GID to the mailserver volumes instead of UID/GID:5000

### DISCUSSION THREAD

**Discussion by fawzibr** [docker-mailserver/docker-mailserver#3407] [Q&A]
Just like the subject says I'm installing the mailserver in a synonogy and cant create users/groups with a given uid/gid. Is there a way to to tell the mailserver to use something other than uid/gid 5000:5000?

**Comment by polarathene** [comment]
Not that I'm aware of it, hard-coded by us.

There is ID Mapped mounts which allows you to map the UID/GID values used in a container to something else on a host bind mount volume. I don't think that's arrived for Docker yet, but Podman has it AFAIK. However with NAS the kernel may not be new enough to support that feature :shrug:

You could probably use named volumes, which should work but then you won't as easily be able to browse that content from the host AFAIK.

Is there some sort of security preventing the use of `5000:5000`? You don't need to actually have any user or group with these values created on the host as that's only a friendly mapping of the UID and GID values. If Docker is running as root I'd expect that to just work?

**Reply by fawzibr** [reply]
_`There is ID Mapped mounts which allows you to map the UID/GID values used in a container to something else on a host bind mount volume. I don't think that's arrived for Docker yet, but Podman has it AFAIK. However with NAS the kernel may not be new enough to support that feature`_

I was hoping there was something similar to the port assignment option (-p 80:80) for uid/gid in docker but there is not. Maybe in the future. It would make all these permission problems easier. Having the same problem with docker MSSQL (they use uid 10001) but it seems they have an option for that. Have to check it.

_`Is there some sort of security preventing the use of 5000:5000? You don't need to actually have any user or group with these values created on the host as that's only a friendly mapping of the UID and GID values. If Docker is running as root I'd expect that to just work?`_

It's working on the Synology NAS but if I try to browse those folders with the NAS UI since the user does not exists it get wonky. Is not really a problem since it works, is just one of those OCD things that is not causing problems but still annoys you. Synology does not let you create users/groups with a given uid/gid, just does it sequentially (starting at 1000). 

I could create a bash script to create users until I can create the 5000 one and then delete the others, and be able to sleep at night.  :)

**Reply by polarathene** [reply]
> I was hoping there was something similar to the port assignment option (-p 80:80) for uid/gid in docker but there is not.

There is user-namespace remapping (_I think that enables mapping 5000 to something else?_). Docker rootless containers also do similar if you follow those docs. I've not used the rootless feature much, but the root user became the non-root user, I think it mapped a range offset from guest to host.

> Synology does not let you create users/groups with a given uid/gid, just does it sequentially (starting at 1000)

You should be able to just modify that after the fact? These should exist in `/etc/passwd`, `/etc/groups` and one other file potentially `/etc/shadow` maybe.

Are you creating the users via a UI? You could just use the `adduser` or `useradd` commands in a shell which allow specifying options for UID/GID, not sure if that would sync with any UI frontend offered.

**Comment by fawzibr** [comment]
> There is user-namespace remapping (_I think that enables mapping 5000 to something else?_). Docker rootless containers also do similar if you follow those docs. I've not used the rootless feature much, but the root user became the non-root user, I think it mapped a range offset from guest to host.

I will have to check that.
 
> You should be able to just modify that after the fact? These should exist in `/etc/passwd`, `/etc/groups` and one other file potentially `/etc/shadow` maybe.

Synology recreates /etc/shadow on reboot/upgrades from somehere. Which from what I read removes any user with uid less than 1024. Anyway while looking for a link that explains the problem, found a solution :)
It seems Synology has a different user database that recreates/verifies /etc/passwd . Apparently just have to create the user and update that database.

> Are you creating the users via a UI? You could just use the `adduser` or `useradd` commands in a shell which allow specifying options for UID/GID, not sure if that would sync with any UI frontend offered.

None of that in synology. Has its own synouser/synogroup programs. With no uid/gid. Will test the new method I found. Thanks anyway.

**Reply by polarathene** [reply]
A feature request was raised to customize the default UID/GID to something other than `5000:5000` via ENV: https://github.com/docker-mailserver/docker-mailserver/issues/3548

**UPDATE:** Merged for v13
